app/validators/uri_validator.rb

```diff
@@ -43,6 +51,40 @@ def forbidden_part?(record, uri)
 
   protected
 
+  def validate_scheme(record, attribute, scheme)
+    valid = valid_scheme?(scheme)
+    add_generic_failure_message(record, attribute) unless valid
+    valid
+  end
+
+  def validate_lack_forbidden_parts(record, attribute, uri)
+    invalid = forbidden_part?(record, uri)
+    add_generic_failure_message(record, attribute) if invalid
+    !invalid
+  end
+
+  def validate_host(record, attribute, host)
+    record.errors[attribute] << :blank and return false if host.blank?
+
+    valid_host_size = valid_host_size?(host)
+    valid_host_labels = valid_host_labels?(host)
+    record.errors[attribute] << I18n.t('activerecord.errors.messages.too_long', count: MAX_HOST_SIZE) unless valid_host_size
+    record.errors[attribute] << I18n.t('activerecord.errors.messages.host_label_too_long', count: MAX_LABEL_SIZE) unless valid_host_labels
+    valid_host_size && valid_host_labels
+  end
+
+  def valid_host_size?(host)
```

Comment:
Put this before `valid_host_labels` for a better readability.

Reply:
@josemigallas i do not want to make this method public unnecessarily... in fact, if we can make the other private, even better... but i didn't have time to check if that is possible as i said in the description